resource/store.go

```diff
@@ -154,6 +160,53 @@ func (s *Store) UpdateSubresource(
 	})
 }
 
+// Upsert updates/creates the provided object.
+// Keep in mind that an Upsert will completely overwrite the object,
+// so nil or missing values will be removed, not ignored.
+// It is usually best to use the result of a Get call, change the appropriate values, and then call Update with that.
+// The update will fail if no ResourceVersion is provided, or if the ResourceVersion does not match the current one.
+// It returns the updated/created Object from the storage system.
+func (s *Store) Upsert(ctx context.Context, obj Object) (Object, error) {
+	if obj.StaticMetadata().Kind == "" {
+		return nil, fmt.Errorf("obj.StaticMetadata().Kind must not be empty")
+	}
+	if obj.StaticMetadata().Namespace == "" {
+		return nil, fmt.Errorf("obj.StaticMetadata().Namespace must not be empty")
+	}
+	if obj.StaticMetadata().Name == "" {
+		return nil, fmt.Errorf("obj.StaticMetadata().Name must not be empty")
+	}
+
+	client, err := s.getClient(obj.StaticMetadata().Kind)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.Get(ctx, obj.StaticMetadata().Identifier())
+
+	if err != nil {
+		cast, ok := err.(APIServerResponseError)
+		if !ok {
+			return nil, err
+		} else if cast.StatusCode() != http.StatusNotFound {
+			return nil, err
+		}
+	}
+
+	if resp != nil {
+		return client.Update(ctx, Identifier{
+			Namespace: obj.StaticMetadata().Namespace,
+			Name:      obj.StaticMetadata().Name,
+		}, obj, UpdateOptions{
+			ResourceVersion: obj.CommonMetadata().ResourceVersion,
```

Comment:
@IfSentient @mdefanti this should actually be the version from the response, not the one from the object passed into the Upsert.